### ISSUE TITLE
Saving manifest is very slow.

### ISSUE THREAD

**Issue by hworld** [elixir-lsp/elixir-ls#162]
### Environment

* Elixir & Erlang versions (elixir --version):
Erlang/OTP 22 [erts-10.6.4] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe]
Elixir 1.10.2 (compiled with Erlang/OTP 22)

* Operating system: Fedora 31
* Editor or IDE name (e.g. Emacs/VSCode): VSCode
* If using VSCode, are you using "ElixirLS Fork: Elixir support and debugger" (note that it has "Fork" in the name)? Yes

---

TLDR; Let's change the compression ratio used when going from term to binary.

So, I'm looking into performance a little, particular when making a modification and it has to build/analyse/save. Expect possibly some more issues as I dig stuff up.

It's been taking ~10s for Elixir LS to go from "save elixir file" to "manifest saved." When I've been looking in my logs, though, I've seen:
[Info  - 10:55:02 PM] Compile took 817 milliseconds
[Info  - 10:55:05 PM] [ElixirLS Dialyzer] Analysis finished in 315 milliseconds

So, clearly the time isn't spent compiling or analyzing, but somewhere else. I realized after a little digging that saving the manifest is very slow for me (I am on a Core i9 laptop with an SSD and 32G of ram).

I pulled some timing info on the whole `Manifest.write` function and realized that it's actually doing compression level 9. I tried finding some documentation but can't find the compression library used, I'm guessing it's just gzip. A compression level of 9 is definitely beyond the line of diminishing returns with gzip.

I pulled some timing info to help make a decision on should be done.

```
No Compression - 193MB Manifest File
2131ms
2037ms
2499ms

Compression 1 - 14.4MB Manifest File
2943ms
3104ms
2826ms

Compression 2 - 12.8MB Manifest File
3078ms
2917ms
2913ms

Compression 5 - 9.8MB Manifest File
3507ms
3472ms
3299ms

Compression 7 - 8.0MB Manifest File
5224ms
3878ms
4514ms

Compression 9 - 7.6MB Manifest File
7488ms
6555ms
6426ms
```

Obviously, a 193MB file is just insanely large to be storing to the disk on every save, so it has to be compressed. It seems like level 5 is about where it starts dropping dramatically, without much gains, however I actually think that for this the goal should be pure speed and just use level 1.

Thoughts on this? I can make a pull request changing this, depending on what is decided.

Secondly, I decided to split the timing of the function up between the part that pulls from the ETS tables, and the part that writes the manifest. This is the numbers I saw.

```
[Info  - 11:06:28 PM] [ElixirLS Dialyzer] Build manifest data, took 1197 milliseconds.
[Info  - 11:06:28 PM] [ElixirLS Dialyzer] Writing manifest...
[Info  - 11:06:30 PM] [ElixirLS Dialyzer] Done writing manifest, took 1469 milliseconds.
```

So, you can see that pulling that amount of erlang memory out of ETS is pretty slow, it's nearly half the time of the `write` function. I'm not super great at elixir/erlang, but I'm wondering if maybe a custom genserver would be better for storing this state. This way it can be compressed before sending over the process boundary. Or maybe even better, just have the process that is storing the data do the save of the manifest so that it doesn't need to transfer.

Another question I had is, how important is it to save the manifest every build/analyse? Can it instead queue up a write to happen after X seconds in the future and sort of debounce any calls? It could also save in a small scratch file the modules that have changed in analysis, queue up the manifest write, and then if the application closes before the write it can make sure that it re-analyses those modules on startup.

**Comment by axelson** [comment]
Nice sleuthing! And thanks for running the benchmarks.

> Obviously, a 193MB file is just insanely large to be storing to the disk on every save, so it has to be compressed. It seems like level 5 is about where it starts dropping dramatically, without much gains, however I actually think that for this the goal should be pure speed and just use level 1.
> 
> Thoughts on this? I can make a pull request changing this, depending on what is decided.

Yeah let's go forward with compression level 1

> So, you can see that pulling that amount of erlang memory out of ETS is pretty slow, it's nearly half the time of the write function. I'm not super great at elixir/erlang, but I'm wondering if maybe a custom genserver would be better for storing this state. This way it can be compressed before sending over the process boundary. Or maybe even better, just have the process that is storing the data do the save of the manifest so that it doesn't need to transfer.

Yeah I believe that the same process that is doing the analysis could also build and write the manifest. I might be missing something but I think instead of transferring ownership of the ETS table from the manifest writer task to the `Dialyzer` GenServer instead the `Dialyzer` GenServer could own the ETS table but allow public write access to it and the analyzing, building, and writing could all be done in a single task. Possibly we might be able to remove the use of ETS completely.

> Another question I had is, how important is it to save the manifest every build/analyse? Can it instead queue up a write to happen after X seconds in the future and sort of debounce any calls? It could also save in a small scratch file the modules that have changed in analysis, queue up the manifest write, and then if the application closes before the write it can make sure that it re-analyses those modules on startup.

This partially already happens, but without the delay. If the manifest is still in the process of being written, the process writing the manifest is killed and restarted: https://github.com/elixir-lsp/elixir-ls/blob/6b70e16671435688eddec23db3023d4d7c7607b1/apps/language_server/lib/language_server/dialyzer.ex#L143

There might be some gains in this area as well, but I think it is limited since this portion is already done asynchronously and nothing is waiting on it. So I think we should plan to tackle the first two issues you brought up first, then from there we could look at how we would like to proceed.

So to recap, for now two separate PR's, one to change the compression level and another to reduce the inter-process communication would be the next steps.

**Comment by lukaszsamson** [comment]
It seems that the low hanging fruits have been picked. I'm going to close it for now until there are new ideas.
